src/colmap/controllers/base_controller.h

```diff
@@ -0,0 +1,112 @@
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include "colmap/util/threading.h"
+
+#include <functional>
+#include <unordered_map>
+
+namespace colmap {
+
+// Reimplementation of threading with thread-related functions outside
+// controller Following util/threading.h
+// BaseController that supports templating in ControllerThread
+class BaseController {
+ public:
+  BaseController();
+  virtual ~BaseController() = default;
+
+  // Set callbacks that can be triggered within the main run function.
+  void AddCallback(int id, const std::function<void()>& func);
+
+  // This is the main run function to be implemented by the child class. If you
+  // are looping over data and want to support the pause operation, call
+  // `BlockIfPaused` at appropriate places in the loop. To support the stop
+  // operation, check the `IsStopped` state and early return from this method.
+  virtual void Run() = 0;
+
+  // check if the thread is stopped
+  void SetCheckIfStoppedFunc(const std::function<bool()>& func);
+  bool CheckIfStopped();
+
+ protected:
+  // Register a new callback. Note that only registered callbacks can be
+  // set/reset and called from within the thread. Hence, this method should be
+  // called from the derived thread constructor.
+  void RegisterCallback(int id);
+
+  // Call back to the function with the specified name, if it exists.
+  void Callback(int id) const;
+
+ private:
+  // list of callbacks
+  std::unordered_map<int, std::list<std::function<void()>>> callbacks_;
+  // check_if_stop function
+  std::function<bool()> check_if_stopped_fn_;
+};
+
+// Helper class to create single threads with simple controls
+// Similar usage as ``Thread`` class in util/threading.h except for
+// initialization. e.g.,
+//
+// std::shared_ptr<Controller> controller = std::make_shared<Controller>(args);
+// std::unique_ptr<ControllerThread<Controller>> thread =
+// std::make_unique<ControllerThread<Controller>>(controller);
+//
+//
+template <class Controller>
+class ControllerThread : public Thread {
+  // check if the Controller class is inherited from BaseController
+  static_assert(std::is_base_of<BaseController, Controller>::value,
+                "The controller needs to be inherited from BaseController");
+
+ public:
+  explicit ControllerThread(std::shared_ptr<Controller> controller)
+      : controller_(std::move(controller)) {
+    controller_->SetCheckIfStoppedFunc([&]() { return IsStopped(); });
```

Comment:
AFAIK here you can also write `SetCheckIfStoppedFunc(std::bind(&IsStopped, this))`

Reply:
Yeah but I guess lambda is fine. Do we want to change this? I can do it

Reply:
I don't care much